internal/controller/apiextensions/composite/composition_ptf.go

```diff
@@ -634,30 +635,87 @@ func FunctionIODesired(s *PTFCompositionState) (iov1alpha1.Desired, error) {
 	return iov1alpha1.Desired{Composite: dxr, Resources: dcds}, nil
 }
 
+// A ContainerFunctionRunnerOption modifies the behavior of a
+// ContainerFunctionRunner, typically by mutating the RunFunctionRequest it
+// produces.
+type ContainerFunctionRunnerOption func(ctx context.Context, fn *v1.ContainerFunction, r *fnv1alpha1.RunFunctionRequest) error
+
+// WithInClusterAuth configures a ContainerFunctionRunner to use in-cluster
+// OCI registry authentication. Image pull secrets will be loaded from the
+// supplied namespace and service account.
+func WithInClusterAuth(namespace, serviceAccount string) ContainerFunctionRunnerOption {
+	return func(ctx context.Context, fn *v1.ContainerFunction, r *fnv1alpha1.RunFunctionRequest) error {
+		ko := k8schain.Options{
+			Namespace:          namespace,
+			ServiceAccountName: serviceAccount,
+			ImagePullSecrets:   make([]string, 0, len(fn.ImagePullSecrets)),
+		}
+
+		for i := range fn.ImagePullSecrets {
+			ko.ImagePullSecrets[i] = fn.ImagePullSecrets[i].Name
+		}
+
+		keychain, err := k8schain.NewInCluster(ctx, ko)
```

Comment:
I started thinking about how to unit test this functionality, and now I'm questioning whether:

* We should use `k8schain.NewFromPullSecrets` instead of `k8schain.NewInCluster`.
* Taking an array of `imagePullSecrets` for each function is a good API in the first place.

For context, `k8schain.NewInCluster` attempts to emulate how the Kubelet authenticates to a private registry. It does this by:

1. Creating a new Kubernetes API client that supports _only_ in-cluster authentication.
2. Fetching all of the supplied [service account's image pull secrets](https://kubernetes.io/docs/concepts/containers/images/#specifying-imagepullsecrets-on-a-pod).
3. Fetching alll of the explicitly named secrets (from within the supplied namespace).

It chains this Kubernetes-secret based authentication with several other credential sources - namely the Docker config file, and credential helpers for GKE, EKS, and AKS.

I believe it would be ideal to fetch any secrets using the same Kubernetes client as the controller manager. We could do this by fetching the secrets in advance, then using `k8schain.NewFromPullSecrets`. This would bring some minor improvements - we'd share client caches, and would be able to load secrets even if Crossplane wasn't running "in-cluster" (i.e. wasn't running as a pod in the same cluster as its API server). I believe it would also make the functionality easier to test, because we'd be injecting secrets rather than having the keychain create and try to use its own Kubernetes client.

This got me thinking about whether `imagePullSecrets` is the best API for us. Specifically what I wonder is:

* How often do folks actually want to specify secrets on a per-function basis? Would just using the secrets associated with the Crossplane pod's service account be sufficient?
* If we do want to keep specifying secrets on a per-function basis, should we continue to assume those secrets will exist in the same namespace as the Crossplane pod? Elsewhere when we reference secrets from a cluster-scoped resource we typically require the namespace of each secret to be specified explicitly - e.g. in an MR or XR's `writeConnectionSecretToRef`.
* Could we eventually want to load image pull secrets from some other secret store (e.g. Vault)?

I think the key reason to stick with `imagePullSecrets` is that we've already committed to it in the package manager API (as `packagePullSecrets`), though all my thoughts above apply to that API too.

Reply:
> How often do folks actually want to specify secrets on a per-function basis? Would just using the secrets associated with the Crossplane pod's service account be sufficient?

I suppose one approach here would be to not add the new `imagePullSecrets` field (or rather, remove it from the master branch before it gets released for the first time with v1.13) and instead support only service-account secrets for pulling functions. We could always add support for specifying secrets at the per-function level if/when there was demand for it.

Reply:
> I believe it would be ideal to fetch any secrets using the same Kubernetes client as the controller manager. We could do this by fetching the secrets in advance, then using k8schain.NewFromPullSecrets. This would bring some minor improvements - we'd share client caches, and would be able to load secrets even if Crossplane wasn't running "in-cluster" (i.e. wasn't running as a pod in the same cluster as its API server). I believe it would also make the functionality easier to test, because we'd be injecting secrets rather than having the keychain create and try to use its own Kubernetes client.

This part all sounds good to me 👍🏻 When the initial package pull secrets functionality was introduced the `go-containerregistry` API looked very different and it was somewhat more difficult to do this. I'd love to see these improvements made across both composition functions and the package manager.

> How often do folks actually want to specify secrets on a per-function basis? Would just using the secrets associated with the Crossplane pod's service account be sufficient?

Fairly indifferent to this one -- the rationale for supporting per-package secrets was less about isolation and more about UX around not having to edit the SA. Since composition functions haven't committed to this model, its worth evaluating whether it is necessary here.

> If we do want to keep specifying secrets on a per-function basis, should we continue to assume those secrets will exist in the same namespace as the Crossplane pod? Elsewhere when we reference secrets from a cluster-scoped resource we typically require the namespace of each secret to be specified explicitly - e.g. in an MR or XR's writeConnectionSecretToRef.

This is arguably an unnecessary restriction, but packages and functions are a little different than other resources that reference secrets in that they are more about configuring how the "crossplane system" works. I haven't heard anyone complain about not being able to put secrets elsewhere, but that doesn't mean folks wouldn't prefer it.

> Could we eventually want to load image pull secrets from some other secret store (e.g. Vault)?

This sounds awesome -- I would highly endorse it.

Reply:
> > If we do want to keep specifying secrets on a per-function basis, should we continue to assume those secrets will exist in the same namespace as the Crossplane pod? Elsewhere when we reference secrets from a cluster-scoped resource we typically require the namespace of each secret to be specified explicitly - e.g. in an MR or XR's writeConnectionSecretToRef.
> 
> This is arguably an unnecessary restriction, but packages and functions are a little different than other resources that reference secrets in that they are more about configuring how the "crossplane system" works. I haven't heard anyone complain about not being able to put secrets elsewhere, but that doesn't mean folks wouldn't prefer it.

I have some concerns around multi-tenancy where a Crossplane instance is shared with different teams and the need for editing a shared XP ServiceAccount. However, for composition functions, it definitely makes sense to at least start with this, and consider extending it if needed.

In general, I do like the clarification we would get from categorizing these as _"about configuring how the "crossplane system" works"_ and providing a consistent experience.

> Could we eventually want to load image pull secrets from some other secret store (e.g. Vault)?

This sounds great, and the proposed changes here would prepare the code for this. However, not sure how users would configure this if we expect the pull secrets to be provided via the ServiceAccount. Wouldn't this require us to have an API for this similar to what we would like to drop now?

Reply:
> However, not sure how users would configure this if we expect the pull secrets to be provided via the ServiceAccount.

Right. To be clear, I agree with your statement that we'd need to introduce some new API to support loading secrets from a secret store. That would just become a "then" problem rather than a "now" problem. 😄 

> Wouldn't this require us to have an API for this similar to what we would like to drop now?

Yes I think it would, but the key word is _similar_.

If we add `imagePullSecrets` today as an array of `LocalObjectReference` it's possible we might find the API limiting tomorrow when we want to support loading these secrets from a different secret store. That said:

* Right now this is an alpha API, so we're not making that big a commitment.
* The current `publishConnectionDetailsTo` object (w/ `configRef`, `metadata`, and `name`) is actually backward compatible with `LocalObjectReference`.

Reply:
> That said

I talked myself back around to just leaving `imagePullSecrets` in for now, for the reasons in my most recent comment. 😄